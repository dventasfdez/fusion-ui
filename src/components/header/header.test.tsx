import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import renderer from 'react-test-renderer';
import Header, {HeaderItem, HeaderLogo, IHeaderProps} from './header';
import {fireEvent, render} from '@testing-library/react';
jest.mock('react-dom', () => ({
  ...jest.requireActual('react-dom'),
  createPortal: (node: any) => node,
}));

const _langs = [
  {id: '1', label: 'en-us', selected: true},
  {id: '2', label: 'es-ES'},
  {id: '3', label: 'en-uk'},
  {id: '4', label: 'it-IT'},
];
const _contact = {label: 'contact', href: '#'};
const HeaderTest = (props: IHeaderProps) => (
  <Header {...props}>
    <HeaderLogo>
      <img
        src="https://upload.wikimedia.org/wikipedia/commons/thumb/9/92/Adecco_logo.svg/1200px-Adecco_logo.svg.png"
        alt="logo"
      />
    </HeaderLogo>
    <HeaderItem data-testid="header-test-item" selected>
      Menu item
    </HeaderItem>
    <HeaderItem
      data-testid="header-test-item-with-options"
      options={[
        {id: '1', label: 'Option', href: '#'},
        {id: '2', label: 'Option', href: '#'},
        {id: '3', label: 'Option'},
      ]}
    >
      Menu item
    </HeaderItem>
    <HeaderItem
      key="item-3"
      data-testid="header-test-item-with-option"
      optionsDivider
      options={[
        {
          id: '1',
          label: 'Option',
          onClick: () => {
            return;
          },
        },
      ]}
    >
      Menu item
    </HeaderItem>
  </Header>
);

describe('Header desktop', () => {
  test('Component renders and match snapshot', () => {
    const component = renderer.create(<HeaderTest />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('Component renders with language selector and match snapshot', () => {
    const component = renderer.create(<HeaderTest languages={_langs} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('Component renders with language selector and contact link and match snapshot', () => {
    const component = renderer.create(<HeaderTest languages={_langs} contact={_contact} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('Select language in header and click in other item (close overflow menu)', () => {
    const onChangeLanguageMock = jest.fn();
    const onSearchMock = jest.fn();
    const {container, getByTestId} = render(
      <HeaderTest
        data-testid="header-test"
        languages={_langs}
        onClickLanguage={onChangeLanguageMock}
        onSearch={onSearchMock}
      />
    );
    const languagesBtn = getByTestId('header-test-languages-dropdown-btn');
    if (languagesBtn) fireEvent.click(languagesBtn);

    const languageItemBtn = getByTestId('header-test-language-2');
    if (languageItemBtn) fireEvent.click(languageItemBtn);
    expect(onChangeLanguageMock).toBeCalledWith('3');

    const option1 = getByTestId('header-test-item-with-options');
    if (option1) fireEvent.click(option1);

    expect(container.getElementsByClassName('overflow-menu').length).toBe(1);
    const option2 = getByTestId('header-test-item');
    if (option2) fireEvent.click(option2);
    const itemOpt = getByTestId('header-test-item-with-option');
    if (itemOpt) fireEvent.click(itemOpt);
    const opt = getByTestId('header-test-item-with-option-overflow-menu-item-0');
    if (opt) fireEvent.click(opt);
  });
});

describe('Header mobile version', () => {
  beforeEach(() => {
    window.innerWidth = 400;
    window.dispatchEvent(new Event('resize'));
  });

  test('Component renders mobile and match snapshot', () => {
    const component = renderer.create(<HeaderTest />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('Component renders mobile with contact selector and match snapshot', () => {
    const component = renderer.create(<HeaderTest contact={_contact} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('Select language in header', () => {
    const onChangeLanguageMock = jest.fn();
    const onSearchMock = jest.fn();
    const {getByTestId} = render(
      <HeaderTest
        data-testid="header-test"
        languages={_langs}
        onClickLanguage={onChangeLanguageMock}
        onSearch={onSearchMock}
      />
    );
    const itemsBtn = getByTestId('header-test-items-dropdown-btn');

    if (itemsBtn) fireEvent.click(itemsBtn);
    const languagesBtn = getByTestId('header-test-languages-dropdown-btn');

    if (languagesBtn) fireEvent.click(languagesBtn);

    const languageItemBtn = getByTestId('header-test-language-2');
    if (languageItemBtn) fireEvent.click(languageItemBtn);
    expect(onChangeLanguageMock).toBeCalledWith('3');
  });
  test('Click in item (close language selector) in header', () => {
    const onChangeLanguageMock = jest.fn();
    const onSearchMock = jest.fn();
    const {container, getByTestId} = render(
      <HeaderTest
        data-testid="header-test"
        languages={_langs}
        onClickLanguage={onChangeLanguageMock}
        onSearch={onSearchMock}
      />
    );
    const itemsBtn = getByTestId('header-test-items-dropdown-btn');

    if (itemsBtn) fireEvent.click(itemsBtn);
    const languagesBtn = getByTestId('header-test-languages-dropdown-btn');

    if (languagesBtn) fireEvent.click(languagesBtn);

    expect(getByTestId('header-test-language-2')).toBeDefined();

    const itemOptions = getByTestId('header-test-item-with-options');
    if (itemOptions) fireEvent.click(itemOptions);

    expect(container.getElementsByClassName('tree-element-item').length).toBe(3);

    const option = getByTestId('header-test-item-with-options-option-1');
    if (option) fireEvent.click(option);
  });

  test('Click in item without href', () => {
    const onChangeLanguageMock = jest.fn();
    const onSearchMock = jest.fn();
    const {getByTestId} = render(
      <HeaderTest
        data-testid="header-test"
        languages={_langs}
        onClickLanguage={onChangeLanguageMock}
        onSearch={onSearchMock}
      />
    );
    const itemsBtn = getByTestId('header-test-items-dropdown-btn');

    if (itemsBtn) fireEvent.click(itemsBtn);
    const itemOptions = getByTestId('header-test-item-with-options');
    if (itemOptions) fireEvent.click(itemOptions);
    const option = getByTestId('header-test-item-with-options-option-2');
    if (option) fireEvent.click(option);
    expect(itemsBtn).toBeDefined();
  });
});
