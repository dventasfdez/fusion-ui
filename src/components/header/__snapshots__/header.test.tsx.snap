// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header desktop Component renders and match snapshot 1`] = `
<div
  className="header "
>
  <img
    alt="logo"
    className="header-logo  "
    src="https://upload.wikimedia.org/wikipedia/commons/thumb/9/92/stepone_logo.svg/1200px-stepone_logo.svg.png"
  />
  <div
    className="tab-list"
  >
    <button
      className="tab-list-item active "
      data-testid="header-test-item"
      onClick={[Function]}
      type="button"
    >
      Menu item
    </button>
    <button
      className="tab-list-item  "
      data-testid="header-test-item-with-options"
      onClick={[Function]}
      type="button"
    >
      Menu item
    </button>
    <button
      className="tab-list-item  "
      data-testid="header-test-item-with-option"
      onClick={[Function]}
      type="button"
    >
      Menu item
    </button>
  </div>
</div>
`;

exports[`Header desktop Component renders with language selector and contact link and match snapshot 1`] = `
<div
  className="header "
  style={
    Object {
      "display": "block",
    }
  }
>
  <div
    className="header-top"
  >
    <a
      className="header-contact link_small"
      href="#"
    >
      contact
    </a>
    <div
      className="dropdown header-languages"
      data-show={false}
    >
      <div
        className="dropdown-button "
        onClick={[Function]}
      >
        en-us
        <span
          className="material-icons"
        >
          expand_more
        </span>
      </div>
    </div>
  </div>
  <div
    className="header-bottom"
  >
    <img
      alt="logo"
      className="header-logo  "
      src="https://upload.wikimedia.org/wikipedia/commons/thumb/9/92/stepone_logo.svg/1200px-stepone_logo.svg.png"
    />
    <div
      className="tab-list"
    >
      <button
        className="tab-list-item active "
        data-testid="header-test-item"
        onClick={[Function]}
        type="button"
      >
        Menu item
      </button>
      <button
        className="tab-list-item  "
        data-testid="header-test-item-with-options"
        onClick={[Function]}
        type="button"
      >
        Menu item
      </button>
      <button
        className="tab-list-item  "
        data-testid="header-test-item-with-option"
        onClick={[Function]}
        type="button"
      >
        Menu item
      </button>
    </div>
  </div>
</div>
`;

exports[`Header desktop Component renders with language selector and match snapshot 1`] = `
<div
  className="header "
>
  <img
    alt="logo"
    className="header-logo  "
    src="https://upload.wikimedia.org/wikipedia/commons/thumb/9/92/stepone_logo.svg/1200px-stepone_logo.svg.png"
  />
  <div
    className="tab-list"
  >
    <button
      className="tab-list-item active "
      data-testid="header-test-item"
      onClick={[Function]}
      type="button"
    >
      Menu item
    </button>
    <button
      className="tab-list-item  "
      data-testid="header-test-item-with-options"
      onClick={[Function]}
      type="button"
    >
      Menu item
    </button>
    <button
      className="tab-list-item  "
      data-testid="header-test-item-with-option"
      onClick={[Function]}
      type="button"
    >
      Menu item
    </button>
  </div>
  <div
    className="dropdown header-languages"
    data-show={false}
  >
    <div
      className="dropdown-button "
      onClick={[Function]}
    >
      en-us
      <span
        className="material-icons"
      >
        expand_more
      </span>
    </div>
  </div>
</div>
`;

exports[`Header mobile version Component renders mobile and match snapshot 1`] = `
<div
  className="header "
>
  <img
    alt="logo"
    className="header-logo  "
    src="https://upload.wikimedia.org/wikipedia/commons/thumb/9/92/stepone_logo.svg/1200px-stepone_logo.svg.png"
  />
  <div
    className="header-content"
  >
    <div
      className="dropdown header-items-dropdown"
      data-show={false}
    >
      <div
        className="dropdown-button "
        onClick={[Function]}
      >
        <span
          className="material-icons"
        >
          menu
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`Header mobile version Component renders mobile with contact selector and match snapshot 1`] = `
<div
  className="header "
  style={
    Object {
      "display": "block",
    }
  }
>
  <div
    className="header-top"
  >
    <a
      className="header-contact link_small"
      href="#"
    >
      contact
    </a>
  </div>
  <div
    className="header-bottom"
  >
    <img
      alt="logo"
      className="header-logo  "
      src="https://upload.wikimedia.org/wikipedia/commons/thumb/9/92/stepone_logo.svg/1200px-stepone_logo.svg.png"
    />
    <div
      className="tab-list"
    >
      <button
        className="tab-list-item active "
        data-testid="header-test-item"
        onClick={[Function]}
        type="button"
      >
        Menu item
      </button>
      <button
        className="tab-list-item  "
        data-testid="header-test-item-with-options"
        onClick={[Function]}
        type="button"
      >
        Menu item
      </button>
      <button
        className="tab-list-item  "
        data-testid="header-test-item-with-option"
        onClick={[Function]}
        type="button"
      >
        Menu item
      </button>
    </div>
  </div>
</div>
`;
